Host Orchestration - Renderability checks for templates

stejskalleos opened the pull request:
The new `Orchestration::Templates` class verifies the host's readiness for the provisioning process and checks that all default os templates are renderable.

The `Orchestration::TFTP` was handling this use case partially, only for ['PXELinux', 'PXEGrub2', 'PXEGrub', 'iPXE'] kinds, leaving other template kinds from the list.

**How to reproduce the issue**

- Foreman + Smart Proxy, provisioning setup
- Assign PXELinux, PXEGrub2, and Kickstart Default templates to OS
- Ensure the Kickstart Default cannot be rendered (for example, with invalid ruby syntax).
- Provision a host

Without this PR, the host is provisioning, and the process is heading to later stages, where it will inevitably fail.

With this PR, the template render checks are the top 1 priority, so we don't waste time with other orchestration steps when the templates are invalid.

stejskalleos committed:
Fixes #38422 - Host Orchestration - Renderability checks for templates

The new `Orchestration::Templates` class verifies the host's readiness
for the provisioning process and checks that templates are renderable.

The `Orchestration::TFTP` was handling this use case partially,
only for ['PXELinux', 'PXEGrub2', 'PXEGrub', 'iPXE'] kinds,
leaving other template kinds from the list.

ShimShtein commented on app/models/concerns/orchestration/templates.rb:
```diff
@@ -0,0 +1,38 @@
+    logger.debug "Scheduling render checks of provisioning templates for #{self}"
+
+    # Filter out templates that are already rendered in the tftp orchestration
+    templates = os.os_default_templates
```
We don't use all the default templates in all the provisioning scenarios. We need to be more specific for a single host: which templates it is going to render (be it default or something that is selected specifically) and test only those

stejskalleos replied:
How do I know which templates will be used? If a template is assigned to OS, we should assume it will be used, right?

stejskalleos replied:
Thinking: We can skip host_init_config and registration for the provisioning process, but others?

ShimShtein replied:
You can use methods defined in https://github.com/theforeman/foreman/blob/develop/app/models/concerns/hostext/operating_system.rb.
Ideally you need to modify `templates_used` to include the selected provisioning method and then it will calculate the templates that will be used for that specific host.

stejskalleos replied:
Digging into the possible solution, checking all templates that are assigned to the OS doesn't make sense; the process should not fail on an invalid cloud-init template when the user is trying to provide a host over a local network.

Checking the status of rendered templates that are selected based on the provisioning method is more difficult than I thought.
We have to filter templates:
* By bootloader
* By the OS (Kickstart or whatever Debian is using)
* By the method (For image mode, the user data, or the cloud init)

This can introduce a lot of complexity, with potential buggy behavior in some combinations that we didn't expect or were unaware of.

More and more I'm inclined to have just some kind of information feature, without any impact on the provisioning process, like in the https://github.com/theforeman/foreman/pull/10541

stejskalleos commented:
Update,
It turned out that I had to implement a method for the PXE loader and templates, see `template_for_pxe_loader`.

ShimShtein commented on app/models/concerns/orchestration/templates.rb:
```diff
@@ -0,0 +1,33 @@
+
+    logger.debug "Scheduling render checks of provisioning templates for #{self}"
+
+    templates = [template_for_pxe_loader, provisioning_template].compact
```
Shoudn't it be 
```suggestion
    templates = [template_for_pxe_loader, available_template_kinds(provisioning).values].compact
```
We're not only looking for a provisioning template. We are looking for the whole set of templates that are relevant to this host and its provisioning type. BTW the provisioning kind is kept as a separate parameter in the controller: https://github.com/theforeman/foreman/blob/c338bc83b002c92c4efb95724a9bb86bda47fcc9/app/controllers/hosts_controller.rb#L620
Make sure you're passing the kind properly from the UI

stejskalleos replied:
> We are looking for the whole set of templates that are relevant to this host and its provisioning type. 

Do we? I want to check templates only relevant to the actual (selected) provisioning method. I don't want to check all template kinds assigned to the host.

stejskalleos replied:
The `available_template_kinds` returns templates for all template kinds assigned to the host, unless it's in image mode, [see here](https://github.com/theforeman/foreman/blob/develop/app/models/concerns/hostext/operating_system.rb#L34).

That's what I don't want to do. I don't care about the render status of the iPXE template when I don't do iPXE provisioning.

ShimShtein replied:
While it's better than nothing, can we test more templates? 
For example in image mode, we need to render more than one template, also in PXE use cases you have the pxe template that invokes the installer and then it is changed to the local disk template.

stejskalleos replied:
The problem is determining which templates to check for every possible provisioning combination correctly.

ekohl replied:
We always render all TFTP methods:
https://github.com/theforeman/foreman/blob/a3962dfdc385c9a3786062ae404b325cad8e89e4/app/models/concerns/orchestration/tftp.rb#L37-L47

I would not include it here because the TFTP step already has all the logic and done fairly early. Everything is rolled back anyway if that fails. Effectively reducing it to:
```suggestion
    templates = [provisioning_template]
```

stejskalleos commented:
New behavior
* Ignoring templates that are rendered in the TFTP task
* For provisioning method `build`, check the assigned `provisioning_template`
* For provisioning method `image`, check either the `user_data` template or the `finish` template

This should be sufficient for the template coverage. 

@ekohl @ShimShtein If you're ok with this behavior, I will add tests.

ekohl commented on app/models/concerns/orchestration/templates.rb:
```diff
@@ -0,0 +1,44 @@
+                  provisioning_template
+                when 'image'
+                  cr = compute_resource
+                  uuid = compute_attributes[cr.image_param_name]
```
I'd be tempted to be defensive and also `return unless uuid`

ShimShtein reviewed (dismissed):
I'll leave the final ACK to @ekohl 
From my point of view it's better than nothing, so I would ACK this.

stejskalleos commented:
Rebased & added tests

ekohl commented on app/models/concerns/orchestration/templates.rb:
```diff
@@ -0,0 +1,45 @@
+                  when 'build'
+                    provisioning_template
+                  when 'image'
+                    image = compute_resource.images.find_by_uuid(compute_attributes[compute_resource.image_param_name])
```
I'm looking at https://github.com/theforeman/foreman/blob/770e8cc3527a45a76358bf529f17dab63145fdf9/app/models/concerns/hostext/operating_system.rb#L33-L45 and wonder if we also need to be defensive here. Is there a chance `compute_resource` is `nil`? Do we need to consider finding the CR via the hostgroup? Can `images` be `nil` or `blank?`?

I suspect that there it's much more defensive because it needs to support the host form where the entity may not be saved yet.

Thinking more: the `user_data` may not be so relevant because it's rendered in the compute orchestration:
https://github.com/theforeman/foreman/blob/770e8cc3527a45a76358bf529f17dab63145fdf9/app/models/concerns/orchestration/compute.rb#L64-L67

That, like TFTP, should have direct feedback. It also has priority 2 so it's one of the very first steps it takes.

Would it be sufficient to only render the finish template and consider that good enough?

ekohl commented on app/models/concerns/orchestration/templates.rb:
```diff
@@ -0,0 +1,45 @@
+
+    logger.debug "Scheduling render checks of template for #{self}"
+
+    queue.create name: _("Check renderability of '%{name}' template.") % { name: template.name },
```
Nit: I think in English it's preferred write it like this, but looking at @Lennonka.
```suggestion
    queue.create name: _("Check renderability of template '%{name}'.") % { name: template.name },
```

Lennonka replied:
I agree with Ewoud's suggestions.

ekohl commented on app/models/concerns/orchestration/templates.rb:
```diff
@@ -0,0 +1,45 @@
+    true
+  rescue => e
+    Foreman::Logging.exception("Error while rendering '#{template.name}' template", e)
+    failure _("Failed to render '%{t}' template, error: %{e}") % { t: template.name, e: e }
```
Same as above
```suggestion
    failure _("Failed to render template '%{t}', error: %{e}") % { t: template.name, e: e }
```

ekohl commented on test/unit/orchestration/templates_test.rb:
```diff
@@ -0,0 +1,30 @@
+  end
+
+  test 'with invalid provisioning template' do
+    template = FactoryBot.build(:provisioning_template, template: "<%= I WILL RAISE AN ERROR %>")
```
I love this template text :)

It reminded me of https://github.com/voxpupuli/puppet-syntax/commit/ceb817621b10c621df6d3c665060cfe4ed44fc0f#diff-912e42831b0b6555ff7a04b9089b9451a35425673939995ea6282ecd780c20fb because in Ruby 3.4 `This is not valid Ruby` suddenly became syntactically valid.

I just verified it and on a plain Ruby 3.4 I see
```console
$ ruby
I WILL RAISE AN ERROR
-:1:in '<main>': uninitialized constant ERROR (NameError)
Did you mean?  IOError
               Errno
```

So as long as those constants aren't defined, this will work. However, you can make it a bit more invalid:
```suggestion
    template = FactoryBot.build(:provisioning_template, template: "<%= I WILL RAISE AN ERROR( %>")
```

stejskalleos commented:
* Fixed error messages
* Updated tests
* `template` method is more defensive, checking only `provision` and `finish` templates, with comments explaining the choices

ekohl commented on app/models/concerns/orchestration/templates.rb:
```diff
@@ -0,0 +1,50 @@
+                    uuid = compute_attributes[compute_resource.image_param_name]
+                    image = compute_resource.images.find_by_uuid(uuid)
+
+                    return unless image
```
Now that the user data is gone, this appears unused. Can we now render the finish template regardless of all this data compute resource related data?

stejskalleos replied:
The issue is, if the image is not found or assigned, should we check the finish template?

ShimShtein replied:
I would second Ewoud's argument. This feels unused and also exposes too much compute resources logic to this concern.
Maybe we car reuse the `template_kinds` method from the [operating_system](https://github.com/theforeman/foreman/blob/develop/app/models/concerns/hostext/operating_system.rb) concern to get the available template types and return the `finish` template in case it appears in the template types output.

ekohl replied:
@ShimShtein yes, that would be the correct approach IMHO. I think it will be something like:
```ruby
def template_kind
  case provision_method
  when 'build'
    # Bootloader configuration templates are rendered in the tftp.rb orchestration
    'provision'
  when 'image'
    template_kind('image').first.name
  else
    # TODO: can this happen? bootdisk?
  end
end

def template
  @template ||= provisioning_template(kind: template_kind)
end
```

ekohl reviewed (commented):
Sorry for the string of reviews. It's also the typical "grow while doing" thing and I'm also learning on the fly
